csrc/alias_analysis.cpp

```diff
@@ -170,15 +174,108 @@ std::optional<Layout> mapInLayoutToOutRoot(
+           std::views::reverse) {
+    if (auto* split = dynamic_cast<Split*>(transform)) {
+      // When split->outer() is parallelized and split->inner() is serial, we
+      // remove split->outer() regardless of its position and replace
+      // split->inner() with split->in(). This way, even when split->outer() is
+      // not adjacent to split->inner() (e.g. when it's outermost), we can
+      // still undo the split.
+      //
+      // Several other cases that I haven't implemented for simplicity.
+      //
+      // When split->outer() is serial and split->inner() is parallelized, we
+      // could remove split->inner() and replace split->outer() with
+      // split->in() regardless of split->inner()'s position.
+      //
+      // When split->outer() and split->inner() are both parallelized, we could
+      // replace either of them with split->in() and remove the other.
+      NVF_ERROR(!split->inner()->isParallelized());
+
+      const auto [outer_contiguity, next_i] =
+          allocation_to_contiguity.erase(split->outer());
+      if (!split->outer()->isParallelized()) {
+        // Check adjacency only if split->outer() is parallelized.
+        if (next_i == allocation_to_contiguity.end() ||
+            next_i->first != split->inner()) {
+          return std::nullopt;
+        }
+      }
+      const auto [inner_contiguity, merge_i] =
+          allocation_to_contiguity.erase(split->inner());
+      const auto [mergeable, contiguity] = mergeContiguity(
+          split->outer()->hasExpandedExtent(),
+          outer_contiguity,
+          split->inner()->hasExpandedExtent(),
+          inner_contiguity);
+      if (!mergeable) {
+        return std::nullopt;
+      }
+      allocation_to_contiguity.insert(merge_i, split->in(), contiguity);
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  Layout layout;
+  for (auto&& [alloc_id, contiguity] : allocation_to_contiguity) {
+    layout.allocation_domain.push_back(alloc_id);
+    layout.contiguity.push_back(contiguity);
+  }
+  NVF_ERROR(
+      std::is_permutation(
+          logical.begin(),
+          logical.end(),
+          layout.allocation_domain.begin(),
+          layout.allocation_domain.end()),
+      "This indicates that logical and allocation are not connected via "
+      "transforms. This is most often caused by forgetting to concretize "
+      "a fusion with dynamic reshapes.");
+  return layout;
+}
+
 bool okToRelayout(
     const TensorView* tv,
     const Layout& new_layout,
     const EmptyAllocationAs empty_allocation_as) {
-  const std::vector<IterDomain*> allocation =
-      (empty_allocation_as == EmptyAllocationAs::kUndetermined
-           ? tv->getAllocationDomain()
-           : tv->getMaybeAllocationDomain());
-  return new_layout.isCompliantWith({allocation, tv->getContiguity()});
+  if (empty_allocation_as == EmptyAllocationAs::kUndetermined &&
+      !tv->hasAllocation()) {
```

Comment:
Why do we need `!tv0->hasAllocation()`? IIUC `isCompliantWith` does check if the existing allocation domain is compatible with the proposed allocation domain. Is that not the case?

Reply:
> IIUC isCompliantWith does check if the existing allocation domain is compatible with the proposed allocation domain. Is that not the case?

Yes. 

> Why do we need !tv0->hasAllocation()?

I don't see how this question is connected with your above statement. But I hope the following breakdown will help you understand: 
* empty_allocation_as_undetermined and tv->hasAllocation: check isCompliantWith, which looks at tv->getMaybeAllocationDomain() and therefore tv->getAllocationDomain()
* empty_allocation_as_undetermined and !tv->hasAllocation: ok to relayout
* empty_allocation_as_logical: check isCompliantWith, which looks at tv->getMaybeAllocation()

Reply:
Ahh my bad. I misunderstood.